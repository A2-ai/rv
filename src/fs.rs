use std::fs;
use std::fs::Metadata;
use std::path::Path;

use filetime::FileTime;
use walkdir::WalkDir;

const MTIME_FILE: &str = "rv.lock";

/// Copy the whole content of a folder to another folder
pub(crate) fn copy_folder(
    from: impl AsRef<Path>,
    to: impl AsRef<Path>,
) -> Result<(), std::io::Error> {
    let from = from.as_ref();
    let to = to.as_ref();

    for entry in WalkDir::new(from) {
        let entry = entry?;
        let path = entry.path();

        let relative = path.strip_prefix(from).expect("walkdir starts with root");
        let out_path = to.join(relative);

        if entry.file_type().is_dir() {
            fs::create_dir_all(&out_path)?;
            continue;
        }

        fs::copy(path, out_path)?;
    }

    Ok(())
}

fn metadata(path: impl AsRef<Path>) -> Result<Metadata, std::io::Error> {
    let path = path.as_ref();
    fs::metadata(path)
}

/// Returns the maximum mtime found in the given folder, looking at all subfolders and
/// following symlinks
/// Taken from cargo crates/cargo-util/src/paths.rs
/// We keep it simple for now and just mtime even if it causes more rebuilds than mtime + hashes
pub(crate) fn mtime_recursive(folder: impl AsRef<Path>) -> Result<FileTime, std::io::Error> {
    let meta = metadata(folder.as_ref())?;
    if !meta.is_file() {
        return Ok(FileTime::from_last_modification_time(&meta));
    }

    // TODO: filter out hidden files/folders?
    let max_mtime = WalkDir::new(folder)
        .follow_links(true)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter_map(|e| {
            if e.path_is_symlink() {
                // Use the mtime of both the symlink and its target, to
                // handle the case where the symlink is modified to a
                // different target.
                let sym_meta = match fs::symlink_metadata(e.path()) {
                    Ok(m) => m,
                    Err(err) => {
                        log::debug!(
                            "failed to determine mtime while fetching symlink metadata of {}: {}",
                            e.path().display(),
                            err
                        );
                        return None;
                    }
                };
                let sym_mtime = FileTime::from_last_modification_time(&sym_meta);
                // Walkdir follows symlinks.
                match e.metadata() {
                    Ok(target_meta) => {
                        let target_mtime = FileTime::from_last_modification_time(&target_meta);
                        Some(sym_mtime.max(target_mtime))
                    }
                    Err(err) => {
                        log::debug!(
                            "failed to determine mtime of symlink target for {}: {}",
                            e.path().display(),
                            err
                        );
                        Some(sym_mtime)
                    }
                }
            } else {
                let meta = match e.metadata() {
                    Ok(m) => m,
                    Err(err) => {
                        log::debug!(
                            "failed to determine mtime while fetching metadata of {}: {}",
                            e.path().display(),
                            err
                        );
                        return None;
                    }
                };
                Some(FileTime::from_last_modification_time(&meta))
            }
        })
        .max() // or_else handles the case where there are no files in the directory.
        .unwrap_or_else(|| FileTime::from_last_modification_time(&meta));
    Ok(max_mtime)
}
