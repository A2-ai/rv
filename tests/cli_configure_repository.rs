use assert_cmd::Command;
use std::fs;
use tempfile::TempDir;

fn create_test_config() -> (TempDir, std::path::PathBuf) {
    let temp_dir = TempDir::new().unwrap();
    let config_path = temp_dir.path().join("rproject.toml");
    
    let config_content = r#"[project]
name = "test"
r_version = "4.4"
repositories = [
    {alias = "posit", url = "https://packagemanager.posit.co/cran/2024-12-16/"}
]
dependencies = [
    "dplyr",
]
"#;
    
    fs::write(&config_path, config_content).unwrap();
    (temp_dir, config_path)
}

#[test]
fn test_configure_repository_add() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "add", "cran",
        "--url", "https://cran.r-project.org",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_add_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_add_config", result);
}

#[test]
fn test_configure_repository_add_with_positioning() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "add", "cran",
        "--url", "https://cran.r-project.org",
        "--first",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_add_first_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_add_first_config", result);
}

#[test]
fn test_configure_repository_replace() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "replace", "posit",
        "--url", "https://packagemanager.posit.co/cran/latest",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_replace_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_replace_config", result);
}

#[test]
fn test_configure_repository_replace_with_new_alias() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "replace", "posit",
        "--alias", "posit-new",
        "--url", "https://packagemanager.posit.co/cran/latest",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_replace_new_alias_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_replace_new_alias_config", result);
}

#[test]
fn test_configure_repository_update_alias() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "update", "posit",
        "--alias", "posit-updated",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_update_alias_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_update_alias_config", result);
}

#[test]
fn test_configure_repository_update_url() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "update", "posit",
        "--url", "https://packagemanager.posit.co/cran/latest",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_update_url_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_update_url_config", result);
}

#[test]
fn test_configure_repository_update_force_source() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "update", "posit",
        "--force-source",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_update_force_source_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_update_force_source_config", result);
}

#[test]
fn test_configure_repository_update_by_url() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "update",
        "--match-url", "https://packagemanager.posit.co/cran/2024-12-16/",
        "--alias", "matched-by-url",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_update_by_url_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_update_by_url_config", result);
}

#[test]
fn test_configure_repository_remove() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "remove", "posit",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_remove_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_remove_config", result);
}

#[test]
fn test_configure_repository_clear() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "clear",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the stdout
    insta::assert_snapshot!("cli_clear_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_clear_config", result);
}

#[test]
fn test_configure_repository_json_output() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "--json",
        "configure", "repository", "add", "cran",
        "--url", "https://cran.r-project.org",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(output.status.success());
    
    // Snapshot the JSON stdout
    insta::assert_snapshot!("cli_json_output_stdout", String::from_utf8_lossy(&output.stdout));
    
    // Snapshot the resulting config
    let result = fs::read_to_string(&config_path).unwrap();
    insta::assert_snapshot!("cli_json_output_config", result);
}

#[test]
fn test_configure_repository_error_missing_alias() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "update",
        "--url", "https://example.com",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(!output.status.success());
    
    // Snapshot the stderr
    insta::assert_snapshot!("cli_error_missing_alias_stderr", String::from_utf8_lossy(&output.stderr));
}

#[test]
fn test_configure_repository_error_nonexistent_alias() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "update", "nonexistent",
        "--url", "https://example.com",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(!output.status.success());
    
    // Verify error contains expected message without path-specific part
    let stderr_str = String::from_utf8_lossy(&output.stderr);
    assert!(stderr_str.contains("Alias not found: nonexistent"));
    
    // Snapshot just the essential error part (removing dynamic paths)
    insta::assert_snapshot!("cli_error_nonexistent_alias_stderr", "Failed to configure repository\n\nCaused by:\n    Alias not found: nonexistent");
}

#[test]
fn test_configure_repository_conflict_flags() {
    let (_temp_dir, config_path) = create_test_config();
    
    let mut cmd = Command::cargo_bin("rv").unwrap();
    cmd.args(&[
        "configure", "repository", "add", "cran",
        "--url", "https://cran.r-project.org",
        "--first", "--last",
        "--config-file", config_path.to_str().unwrap()
    ]);
    
    let output = cmd.output().unwrap();
    assert!(!output.status.success());
    
    // Normalize binary name for cross-platform compatibility (rv vs rv.exe)
    let stderr_str = String::from_utf8_lossy(&output.stderr).replace("rv.exe", "rv");
    insta::assert_snapshot!("cli_conflict_flags_stderr", stderr_str);
}