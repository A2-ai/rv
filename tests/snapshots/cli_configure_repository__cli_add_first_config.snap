---
source: tests/cli_configure_repository.rs
expression: result
---
[project]
name = "test"
r_version = "4.4"
repositories = [
    { alias = "cran", url = "https://cran.r-project.org/" },
    {alias = "posit", url = "https://packagemanager.posit.co/cran/2024-12-16/"},
]
dependencies = [
    "dplyr",
]
